### PR TITLE
Adds labels to the Akka HTTP metrics [PLEN-12]

### PR DESCRIPTION
**Pull request by lucb-da**
The basic HTTP labels are automatically added by the metrics support.
The labels for the websockets have to be given to the metrics support, as it has no access to the request itself.

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: observability/akka-http-metrics/src/test/suite/scala/com/daml/metrics/akkahttp/WebSocketMetricsSpec.scala [hunk 1]
```diff
@@ -240,34 +278,47 @@ class AkkaHttpMetricsSpec extends AsyncWordSpec with AkkaBeforeAndAfterAll with
     }
   }
 
-  private val getMessagesReceivedTotalValue: TestMetrics => Long = (metrics: TestMetrics) =>
-    metrics.messagesReceivedTotalValue
+  private val getMessagesReceivedTotalValue: (TestMetrics, Seq[(String, String)]) => Long =
```
Comment (miklos-da) [inline observability/akka-http-metrics/src/test/suite/scala/com/daml/metrics/akkahttp/WebSocketMetricsSpec.scala:281]: To improve readability, can we define a case class (e.g., `LabelKeyValue`) for this tuple, please?

File: observability/akka-http-metrics/src/main/scala/com/daml/metrics/akkahttp/AkkaHttpMetrics.scala [hunk 2]
```diff
@@ -67,6 +61,32 @@ object AkkaHttpMetrics {
 
     }
 
+  // Split method needed to be able to define a new implicit MetricsContext value.
```
[the remark this answers is not in the document]
Comment (lucb-da) [reply observability/akka-http-metrics/src/main/scala/com/daml/metrics/akkahttp/AkkaHttpMetrics.scala:64]: We are in the metrics interceptor. Everything is about metrics in this file :)

File: ledger/metrics/src/main/scala/com/daml/metrics/api/MetricsContext.scala [hunk 3]
```diff
@@ -21,7 +21,10 @@ case class MetricsContext(labels: Map[String, String]) {
 
 object MetricsContext {
 
-  val Empty: MetricsContext = MetricsContext(Map.empty)
+  val Empty: MetricsContext = MetricsContext(Map.empty[String, String])
+
+  def apply(labels: (String, String)*): MetricsContext =
```
Comment (nicu-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/MetricsContext.scala:26]: :+1:

**Review by nicu-da — APPROVED** [review]

File: observability/akka-http-metrics/src/main/scala/com/daml/metrics/akkahttp/LabelExtractor.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+package com.daml.metrics.akkahttp
+
+import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
+
+object LabelExtractor {
+
+  def labelsFromRequest(request: HttpRequest): Seq[(String, String)] = {
+    val baseLabels = Seq[(String, String)](
+      ("http_verb", request.method.name),
```
Comment (miklos-da) [inline observability/akka-http-metrics/src/main/scala/com/daml/metrics/akkahttp/LabelExtractor.scala:12]: I'd strongly suggest defining these magic constants for the labels we generate in a single place and referencing them from there.

**Review by miklos-da — APPROVED** [review]

File: observability/akka-http-metrics/src/test/suite/scala/com/daml/metrics/akkahttp/WebSocketMetricsSpec.scala [hunk 5]
```diff
@@ -41,6 +43,9 @@ class AkkaHttpMetricsSpec extends AsyncWordSpec with AkkaBeforeAndAfterAll with
   private val streamedBinaryMessageSize =
     streamedBinaryMessageData.foldLeft(0L)((acc, bs) => acc + bs.length)
 
+  private val labels = List(("label1", "abc"), ("other", "label2"))
+  private val labelFilter = labels.map { case (key, value) => LabelKeyValue(key, value) }
```
[the remark this answers is not in the document]
Comment (lucb-da) [reply observability/akka-http-metrics/src/test/suite/scala/com/daml/metrics/akkahttp/WebSocketMetricsSpec.scala]: It is now a list of filters, so the name match better the content.

**Review by nicu-da — APPROVED** [review]
lgtm!

**Pull request merged**